### PR TITLE
Documentation on client and EntityDecoder/Encoder

### PR DESCRIPTION
**Pull request by reactormonk**
Still got a few TODOs to fix.

### CODE REVIEW

File: docs/src/main/tut/client.md [hunk 1]
```diff
@@ -109,5 +109,33 @@ server.shutdownNow()
...
+
+However, your function has to consume the body before the returned `Task` exits.
+Don't do this:
+
+```scala
+// will come back to haunt you
+client.get[EntityBody]("some-url")(request => request.body)
+```
+
+Passing it to a `EntityDecoder` is save.
```
Comment (ritschwumm) [inline docs/src/main/tut/client.md]: safe

File: docs/src/main/tut/entity.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
+---
+layout: default
+title: Dealing with Entity Bodies
+---
+
+## Why Entity*
+
+Http4s handles HTTP requests and responses in a streaming fashing. Your service
+will recieve a request after the header has been parsed (ok, not 100%
```
Comment (ritschwumm) [inline docs/src/main/tut/entity.md]: receive

File: docs/src/main/tut/entity.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
+---
+layout: default
+title: Dealing with Entity Bodies
+---
+
+## Why Entity*
+
+Http4s handles HTTP requests and responses in a streaming fashing. Your service
```
Comment (ritschwumm) [inline docs/src/main/tut/entity.md]: fashion

File: docs/src/main/tut/entity.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+will recieve a request after the header has been parsed (ok, not 100%
+streaming), but before the body has been fully recieved. The same applies to the
```
Comment (ritschwumm) [inline docs/src/main/tut/entity.md]: receive

File: docs/src/main/tut/client.md [hunk 5]
```diff
@@ -109,5 +109,33 @@ server.shutdownNow()
...
+```scala
+TODO: Example here
+```
+
+However, your function has to consume the body before the returned `Task` exits.
+Don't do this:
+
+```scala
+// will come back to haunt you
+client.get[EntityBody]("some-url")(request => request.body)
```
Comment (rossabaker) [inline docs/src/main/tut/client.md]: The callback here is `response`, not `request`

File: docs/src/main/tut/client.md [hunk 6]
```diff
@@ -109,5 +109,33 @@ server.shutdownNow()
...
+```
+
+Passing it to a `EntityDecoder` is safe.
+
+```
+client.get[T]("some-url")(request => jsonOf(request.body))
```
Comment (rossabaker) [inline docs/src/main/tut/client.md]: `s/request/response/g`

File: docs/src/main/tut/entity.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+materialized. You don't have to load the full body into memory to submit the
+request either. Taking a look at `Request` and `Response`, both have a body of
+type `EntityBody`, which is simply an alias to `Process[Task, ByteVector]`. To
+understand `Process`, take a look at the [streams-tutorial].
+
+The `EntityDecoder` and `EntityEncoder` help with the streaming nature of the
+data in a http body, and they also have additional logic to deal with media
+types. Not all decoders are streaming, depending on the implementation.
+
+## Working with Media Types
```
Comment (rossabaker) [inline docs/src/main/tut/entity.md]: This paragraph seems a bit scattered to me: it talks about EntityDecoder, and then the headers that EntityEncoder adds, and then back to how to do an EntityDecoder.  I would cover one, then the other.

File: docs/src/main/tut/entity.md [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+types. Because of the implicit conversions, you can also use `(String, String)`
+for a `MediaType`.
+
+### Strictness
+
+By default, the decoders are strict, so you will get an error when the headers
+tell you about a different data type. In some cases, an outside API might be
+broken and give you text/text as `Content-Type`, but it's really json.
+
+TODO: What do?
```
Comment (rossabaker) [inline docs/src/main/tut/entity.md]: Might be a good opportunity to use tut to show a body that decodes non-strictly, and then the error that happens with the strict flag set.

File: docs/src/main/tut/entity.md [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+## Presupplied Encoders/Decoders
+The `EntityEncoder`/`EntityDecoder`s shipped with http4s.
```
Comment (rossabaker) [inline docs/src/main/tut/entity.md]: There is also an Encoder for Twirl templates.

File: docs/src/main/tut/dsl.md [hunk 10]
```diff
@@ -221,3 +221,119 @@ Ok(drip).run
...
+      Try(LocalDate.parse(str)).toOption
+    else
+      None
+  }
+}
+
+def getTemperatureForecast(date: LocalDate): Task[Double] = ???
+
+val dailyWeatherService = HttpService {
+  case request @ GET -> Root / "weather" / "temperature" / LocalDateVar(localDate) =>
```
Comment (rossabaker) [inline docs/src/main/tut/dsl.md]: Good example, but maybe show a request that matches this?

GET("/weather/temperature/2016-11-05")

File: docs/src/main/tut/entity.md [hunk 11]
```diff
@@ -18,23 +18,53 @@ The `EntityDecoder` and `EntityEncoder` help with the streaming nature of the
...
 
-## Working with Media Types
+## Construction and Media Types
 `Entity*`s also encode which media types they correspond to. The
-`EntityDecoders` for json expect `application/json`, and when you encode a body
-with the `EntityEncoder` for json, it appends the `Content-Type:
-application/json` header. To implement this functionality, the constructor
-`EntityDecoder.decodeBy` uses `MediaRange`s, you can pass multiple as needed.
-See the [MediaRange] companion object for ranges, and [MediaType] for specific
+`EntityDecoders` for json expect `application/json`. To implement this
```
Comment (rossabaker) [inline docs/src/main/tut/entity.md]: s/`EntityDecoders`/`EntityDecoder`s/ -- keeps the literal code in quotes and the English inflection out.  (Like you did with `MediaRange`s.)

File: docs/src/main/tut/entity.md [hunk 12]
```diff
@@ -18,23 +18,53 @@ The `EntityDecoder` and `EntityEncoder` help with the streaming nature of the
...
+functionality, the constructor `EntityDecoder.decodeBy` uses `MediaRange`s, you
```
Comment (rossabaker) [inline docs/src/main/tut/entity.md]: s/, you/. You/

File: docs/src/main/tut/entity.md [hunk 13]
```diff
@@ -18,23 +18,53 @@ The `EntityDecoder` and `EntityEncoder` help with the streaming nature of the
...
+can pass multiple as needed. You can also append functionality to an existing
+one via `implicitly[EntityDecoder[T]].map` - however, you can't change the media
```
Comment (rossabaker) [inline docs/src/main/tut/entity.md]: Don't we have an apply method here to summon them?  `EntityDecoder[T]`?  If not, should we?

File: docs/src/main/tut/entity.md [hunk 14]
```diff
@@ -18,23 +18,53 @@ The `EntityDecoder` and `EntityEncoder` help with the streaming nature of the
...
+`Content-Type: application/json` header. You can construct new encoders via
+`EntityEncoder.encodeBy` or reuse an already existing one via
+`implicitly[EntityEncoder[T]].contramap` and `withContentType`.
+
+ See the [MediaRange] companion object for ranges, and [MediaType] for specific
 types. Because of the implicit conversions, you can also use `(String, String)`
 for a `MediaType`.
 
-### Strictness
+By default, decoders ignore content types - it would be rather unpleasant have
```
Comment (rossabaker) [inline docs/src/main/tut/entity.md]: Instead of "by default", maybe something about using `as`?

File: docs/src/main/tut/entity.md [hunk 15]
```diff
@@ -58,6 +88,11 @@ cause conflicts with the builtin Decoders.
 
 - scala-xml: `"org.http4s" %% "http4s-scala-xml" % Http4sVersion`
 
+### Twirl
+If you're working with twirl templates, there's a bridge for that too:
```
Comment (rossabaker) [inline docs/src/main/tut/entity.md]: Maybe link to Twirl?  People may not know what it is.

File: docs/src/main/tut/entity.md (hunk 13 above)
Comment (reactormonk) [inline docs/src/main/tut/entity.md]: There is. I tend to use the `implicitly` because that works everywhere.

**Comment by rossabaker** [comment]
👍 on Travis

**Pull request merged**